### ISSUE TITLE
Incorrect positions in AST and output of -tstats

### ISSUE THREAD

**Issue by cpitclaudel** [EasyCrypt/easycrypt#73]
Hi there,

The -tstats option returns an incorrect range for some commands:

```console
$ printf "require import Bool.\ntime require import Int.\nrequire import IntDiv.\ntime require import Real." > input.ec
$ easycrypt -tstats /dev/stdout -no-eco input.ec
1:0 1:19 0.182414
2:4 2:23 0.000399
1:0 3:21 0.145686
4:4 4:24 0.000499
```

Adding `time` at the beginning of an incorrectly positioned command reliably fixes the issue.  Probably an issue in the lexer?

**Comment by cpitclaudel** [comment]
(Btw, it would be very convenient if the spans that EasyCrypt reports included the `time` prefix, and ideally the final `.`, so that the output of `-tstats` precisely delimits every command in the file — bonus points if it includes `undo` commands as well)

**Comment by strub** [comment]
The `time` prefix and the final `.` should be in the positions range. I don't understand why all the positions are off.

**Comment by cpitclaudel** [comment]
I think it comes from here:

```
global:
| tm=boption(TIME) g=loc(global_action) FINAL
  { { gl_action = g; gl_timed = tm; } }
```

This says that the location on `g` should just be that of the `global_action` (so `tm` isn't included, nor `FINAL`).  There seems to be a bug in the location of `global_action` itself: some actions get the right position, but not all (for example `lemma` and `print` are OK…)

**Comment by cpitclaudel** [comment]
Maybe it's because the rule for `require` starts with an optional value (`nm`)?

```
theory_require:
| nm=prefix(FROM, uident)? REQUIRE ip=import_flag? x=theory_require_1+
    { (nm, x, ip) }
```

**Comment by strub** [comment]
Progressing :)

```
1:0 2:0 0.155827% (-1.0B / [frag -1.0B])
2:0 3:0 0.000525% (-1.0B / [frag -1.0B])
1:0 4:0 0.120391% (-1.0B / [frag -1.0B])
4:0 4:25 0.000471% (-1.0B / [frag -1.0B])
```

**Comment by strub** [comment]
Should be fixed in c07e88032b0dc6ec2a6aa30b6025dbd8fc253515

**Comment by cpitclaudel** [comment]
Thanks!  I see time included in the sentence range now and the period too; that's great :)  But for the third sentence above the start position is still wrong, right?  I'm seeing this:

```
1:0 1:20 0.150206
2:0 2:24 0.000681
1:0 3:22 0.121498
4:0 4:25 0.000246
```

**Comment by strub** [comment]
Yes. This seems related to the handling of positions in parsing rules that
start with an optional symbol. I don’t really know what to do here.

**Comment by cpitclaudel** [comment]
Got it, thanks; I might investigate
